wandb/apis/public.py

```diff
@@ -4354,6 +4358,20 @@ def _use_as(self, use_as):
         self._attrs["_use_as"] = use_as
         return use_as
 
+    @normalize_exceptions
+    def update_aliases(self, add: List[str], remove: List[str]):
```

Comment:
Can you help me understand the difference between this and `art.aliases.append("best")` then calling `art.save()`? Seems like if we are already scoping aliases to the current collection, then this function just augments typically python methods?

Reply:
Nope -- `update_aliases` makes local changes to two new arrays `self._aliases_to_add` and `self._aliases_to_remove`. Then we send those arrays directly to the endpoints `addAliases` and `deleteAliases`.

Reply:
but that is implementation detail correct?

Reply:
like you could also figure out the delta between the starting aliases and the new alias state

Reply:
Hm we need to have some way of storing local alias changes that are different than the starting point. `aliases` is just a list. It sounds like you want `aliases.append(...)` to store these local alias changes for addition, as an example?

Reply:
In either case, we're storing the changes in another data structure so not sure what the proposal is here?

Reply:
From a user perspective, there are now two ways to update aliases: via mutating the .aliases field or calling this new method. I think it would be cleaner to support just the first. When it comes to implementation. We could have two fields: `aliases` and `_original_aliases`. the user manipulates the former, then during save, you use the delta to make the correct call

Reply:
This works if the user does `artifact.aliases = [...]`. However how can we hook into the `append` method on the list?

Reply:
One idea is to create a class that wraps around a list, like `AliasesList` which manages local changes. It would have `aliases` and `_original_aliases` as attributes and list methods as class methods, such as `append`.

Reply:
Just syncing back here: `aliases` will remain a mutable copy of the aliases that the user can modify using Pythonic list ops. We will store a second, frozen copy of the aliases. Then at `save()` time, we'll compute a diff and send accordingly.